android/java/org/chromium/chrome/browser/bookmarks/BraveBookmarkManager.java

```diff
@@ -0,0 +1,163 @@
+    }
+
+    public void setWindow(ActivityWindowAndroid window) {
+        mWindowAndroid = window;
+    }
+
+    @Override
+    public void importBookmarks() {
+        if (mWindowAndroid == null || mContext == null) {
+            return;
+        }
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+            doImportBookmarks();
+        } else {
+            if (mWindowAndroid.hasPermission(Manifest.permission.READ_EXTERNAL_STORAGE)) {
+                doImportBookmarks();
+            } else {
+                mWindowAndroid.requestPermissions(
+                        new String[] {Manifest.permission.READ_EXTERNAL_STORAGE},
+                        (permissions, grantResults) -> {
+                            if (grantResults.length >= 1
+                                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                                doImportBookmarks();
+                            } else {
+                                Log.e("tapan", "else import bookmarks");
+                            }
+                        });
+            }
+        }
+    }
+
+    private void doImportBookmarks() {
+        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
+        intent.setType("text/html");
+        intent.addCategory(Intent.CATEGORY_OPENABLE);
+        if (mWindowAndroid.showIntent(Intent.createChooser(intent,
+                                              mContext.getResources().getString(
+                                                      R.string.import_bookmarks_select_file)),
+                    new WindowAndroid.IntentCallback() {
+                        @Override
+                        public void onIntentCompleted(int resultCode, Intent results) {
+                            if (resultCode == Activity.RESULT_OK && results != null
+                                    && results.getData() != null) {
+                                try {
+                                    Cursor cursor = mContext.getContentResolver().query(
+                                            results.getData(), null, null, null, null);
+                                    int nameIndex =
+                                            cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
+                                    cursor.moveToFirst();
+                                    String name = cursor.getString(nameIndex);
+                                    File file = new File(mContext.getFilesDir(), name);
+                                    try (InputStream inputStream =
+                                                    mContext.getContentResolver().openInputStream(
+                                                            results.getData());
+                                            FileOutputStream outputStream =
+                                                    new FileOutputStream(file)) {
+                                        int maxBufferSize = 1 * 1024 * 1024;
```

Comment:
File work has to be executed in a non UI thread. Check that as an example https://github.com/brave/brave-core/blob/master/android/java/org/chromium/chrome/browser/app/BraveActivity.java#L2060

Reply:
the
```
                                        mBookmarkModel.importBookmarks(
                                                mWindowAndroid, file.getPath());
```
has an execution of a background thread inside, so we can call it in the main thread, but file read and write before it has to be moved to the background thread.